screensaverd.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * Simple screen saver daemon. Turns off the display after idle timeout.
+ * Turns the display on on any event from /dev/input/event* files.
+ * On exit turns the display on.
+ *
+ * Copyright (C) 2015 Jolla Ltd.
+ * Contact: Igor Zhbanov <igor.zhbanov@jolla.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#define _BSD_SOURCE
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <dirent.h>
+#include <signal.h>
+#include <stdarg.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <sys/select.h>
+
+#define UNUSED			__attribute__((unused))
+
+#define DISPLAY_CONTROL		"/sys/class/graphics/fb0/blank"
+#define DEV_INPUT_DIR		"/dev/input"
+#define LOG_NAME		"screensaverd"
+#define EVENT_PREFIX		"event"
+#define MAX_DEVICES		32
+#define DISPLAY_OFF_TIME	25 /* seconds */
+#define BUF_SIZE		4096
+
+sig_atomic_t volatile running = 1;
+
+/* ------------------------------------------------------------------------ */
+
+/* perror() with printf-style format and application name prefix. */
+static void
+perrorf(const char *format, ...)
+{
+	va_list ap;
+
+	va_start(ap, format);
+	printf("[%s] ", LOG_NAME);
+	vprintf(format, ap);
+	printf(": %s\n", strerror(errno));
```

Comment:
The printf() family functions use dynamic memory, character set & floating point conversions, do low level io, etc
And at least AFAIK there is no attempt made to guarantee (in c spec) that errno stays intact, so if it does in one impl it does not mean it is that way in others.

Simple example: errno = 0, rc = fprintf(stdin, "foo\n") -> I get rc=-1, errno=EBADF

Still none of this matters since the current code does not check errno after calling the logging function. But later on someone might assume something and something goes wrong.